src/Umbraco.Web/Models/ImageProcessorImageUrlGenerator.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System.Globalization;
+using System.Text;
+using Umbraco.Core;
+using Umbraco.Core.Models;
+
+namespace Umbraco.Web.Models
+{
+    internal class ImageProcessorImageUrlGenerator : IImageUrlGenerator
+    {
+        public string GetImageUrl(ImageUrlGenerationOptions options)
```

Comment:
I think we should separate several private methods from this method..

Eg:
```csharp
        private void AppendCrop(StringBuilder imageProcessorUrl, ImageUrlGenerationOptions.CropCoordinates crop)
        {
            imageProcessorUrl.Append("?crop=");
            imageProcessorUrl.Append(crop.X1.ToString(CultureInfo.InvariantCulture)).Append(",");
            imageProcessorUrl.Append(crop.Y1.ToString(CultureInfo.InvariantCulture)).Append(",");
            imageProcessorUrl.Append(crop.X2.ToString(CultureInfo.InvariantCulture)).Append(",");
            imageProcessorUrl.Append(crop.Y2.ToString(CultureInfo.InvariantCulture));
            imageProcessorUrl.Append("&cropmode=percentage");
        }

        private void AppendFocalPoint(StringBuilder imageProcessorUrl, ImageUrlGenerationOptions.FocalPointPosition focalPoint)
        {
            imageProcessorUrl.Append("?center=");
            imageProcessorUrl.Append(focalPoint.Top.ToString(CultureInfo.InvariantCulture)).Append(",");
            imageProcessorUrl.Append(focalPoint.Left.ToString(CultureInfo.InvariantCulture));
            imageProcessorUrl.Append("&mode=crop");
        }
```

Reply:
Is any reason why we set up ImageProcessorImageUrlGenerator as internal? I think it should be public to allow for overrides :)

Reply:
I agree with @bielu: make this class public and (most) helper methods protected, so it can be easily overwritten. One _very_ useful example is to add a token to all generated image URLs that can be validated using `ImageProcessingModule.OnProcessQuerystring`, so it's possible to protect against a DoS attack (requesting a lot of images with different parameters that require a lot of server processing power to generate), see: https://github.com/jimbobsquarepants/imageprocessor/issues/49#issuecomment-199337644